js/ui/components/codingGameService.js

```diff
@@ -1,52 +1,152 @@
+            logError(e, 'Error occurred in creating CodingGameServiceProxy');
             return;
         }
 
-        // It doesn't seem possible to return an array of strings here, looks like it has to be an array
-        // of tuples (which contain strings).
-        this._listenForLessonsId =
-            this._service.connect('listening-for-lesson-events', Lang.bind(this, function(proxy, interestingEvents) {
-                let events = interestingEvents.deep_unpack().map(function(i) {
-                    return i[0];
-                });
+        let eventHandlers = {
+            'move-window': Lang.bind(this, this._waitForNextWindowMove),
+            'stop-moving-windows': Lang.bind(this, this._waitForNoWindowsToMove)
+        };
 
-                if (events.indexOf('window-moved') !== -1)
-                    this._waitForNextWindowMove();
+        this._signalHandlers = {};
+        this._listenForEventsId =
+            this._service.connect('listen-for-event', Lang.bind(this, function(proxy, name) {
+                if (eventHandlers[name]) {
+                    this._disconnectHandlersFor(name);
+                    eventHandlers[name]();
+                }
             }));
+        this._stopListeningForEventsId =
+            this._service.connect('stop-listening-for', Lang.bind(this, function(proxy, name) {
+                // In this case, just disconnect any known event handlers, don't
+                // need any custom logic here
+                this._disconnectHandlersFor(name);
+            }));
+
+        this._service.call_currently_listening_for_events(null, Lang.bind(this, function(source, result) {
+            let success, returnValue;
+
+            try {
+                [success, returnValue] = this._service.call_currently_listening_for_events_finish(result);
+            } catch (e) {
+                logError(e);
+                return;
+            }
+
+            returnValue.deep_unpack().map(Lang.bind(this, function(name) {
+                if (eventHandlers[name]) {
+                    this._disconnectHandlersFor(name);
+                    eventHandlers[name]();
+                }
+            }));
+        }));
     },
 
     disable: function() {
-        if (this._listenForLessonsId > 0) {
-            this._service.disconnect(this._listenForLessonsId);
-            this._listenForLessonsId = 0;
+        Object.keys(this._signalHandlers, Lang.bind(this, function(key) {
+            this._disconnectHandlersFor(key);
+        }));
+    },
+
+    _disconnectHandlersFor: function(name) {
+        name = name.replace(/\-/g, '_');
+        if (!this._signalHandlers[name]) {
+            return;
         }
+
+        this._signalHandlers[name].forEach(function(handler) {
+            if (handler.object) {
+                handler.object.disconnect(handler.connection);
+            } else {
+                // Asumme this is a GSource
```

Comment:
Typo: `assume`
But this is not used and feels wrong, so I would simply remove it.